### PR TITLE
Common __repr__ for common types

### PR DESCRIPTION
**Pull request by kernc**
##### Issue
Our types all around.

##### Description of changes
Makes most of our types repr-able, as recommended by [python docs](https://docs.python.org/3/reference/datamodel.html#object.__repr__) and as needed eventually by https://github.com/biolab/orange3/pull/1449.
A continuation of https://github.com/biolab/orange3/pull/1507.

##### Includes
- [X] Code changes
- [X] Tests
- [X] Documentation

This "traits" programming is quite clean. I reckon we shall soon also have types that are `_Copyable` and `_Enumerable`!

Would _really_ appreciate a speedy review!

### CODE REVIEW

**Comment by Ameobea** [comment]
I'm glad that this is finally getting the attention it deserves.  I was not cut out for the level of expertise necessary to make this happen and on top of that got very distracted by other things once summer ended.  

In case my work on code generation could be of any use in this new effort, here's the branch in which I was working (this is other work than my efforts in `__repr__` generation): https://github.com/Ameobea/orange3/tree/code-generation

Sorry I wasn't able to make this happen, and good luck with your efforts.

**Comment by kernc** [comment]
Oh, you did great. If it weren't for you, we likely wouldn't be looking at anything like this at all. :wink:

File: Orange/base.py [hunk 1]
```diff
@@ -262,13 +294,14 @@ def fit(self, X, Y, W=None):
...
-    def __str__(self):
-        return '{} {}'.format(self.name, self.params)
-
     def __repr__(self):
-        return '{}({})'.format(type(self).__name__,
-                               ", ".join("{}={}".format(k, v)
-                                         for k, v in self.params.items()))
+        # In addition to saving values onto self, SklLearners save into params
+        vars = self.__dict__
+        self.__dict__ = dict(vars, **self.params)
```
Comment (astaric) [inline Orange/base.py]: This looks alot like patching the self.__dict__. Any specific reason why not using [patch.object](https://docs.python.org/3/library/unittest.mock.html#patch-object)?

File: Orange/classification/logistic_regression.py [hunk 2]
```diff
@@ -36,7 +36,7 @@ def coefficients(self):
 class LogisticRegressionLearner(SklLearner, _FeatureScorerMixin):
     __wraps__ = skl_linear_model.LogisticRegression
     __returns__ = LogisticRegressionClassifier
-    name = 'logreg'
+    name = 'logit'
```
Comment (astaric) [inline Orange/classification/logistic_regression.py]: Why?

File: Orange/data/filter.py [hunk 3]
```diff
@@ -379,16 +375,11 @@ def __str__(self):
             return "{} is defined".format(column)
         return "invalid operator"
 
-    __repr__ = __str__
-
-
-    # For PyCharm:
-    Equal = NotEqual = Less = LessEqual = Greater = GreaterEqual = 0
```
Comment (astaric) [inline Orange/data/filter.py]: I guess this was here to allow IDEs to suggest these names for code completion. Does this work without the dummy members in the current version of PyCharm?

File: Orange/preprocess/continuize.py [hunk 4]
```diff
@@ -1,12 +1,13 @@
 from Orange.data import ContinuousVariable, Domain
 from Orange.statistics import distribution
+from Orange.util import Reprable
 from .transformation import Identity, Indicator, Indicator1, Normalizer
 from .preprocess import Continuize
 
-__all__ = ["DomainContinuizer", "MultinomialTreatment"]
+__all__ = ["DomainContinuizer"]
```
Comment (astaric) [inline Orange/preprocess/continuize.py]: Why was this removed?

File: Orange/preprocess/preprocess.py [hunk 5]
```diff
@@ -419,7 +502,7 @@ class PreprocessorList:
         A list of preprocessors.
     """
 
-    def __init__(self, preprocessors):
+    def __init__(self, preprocessors=[]):
```
Comment (astaric) [inline Orange/preprocess/preprocess.py]: lint probably complains that having mutable default values is not a good idea.

File: Orange/tests/test_classification.py [hunk 6]
```diff
@@ -8,11 +8,15 @@
 
 import traceback
 import numpy as np
+import Orange
```
Comment (astaric) [inline Orange/tests/test_classification.py]: Is this used anywhere?

File: Orange/widgets/data/owdatasampler.py [hunk 7]
```diff
@@ -219,6 +220,7 @@ def commit(self):
         self.send("Data Sample", sample)
         self.send("Remaining Data", other)
 
+
```
Comment (astaric) [inline Orange/widgets/data/owdatasampler.py]: One empty line is preferred between methods.

File: Orange/widgets/data/owpreprocess.py [hunk 8]
```diff
@@ -762,8 +700,8 @@ def __init__(self, parent=None, **kwargs):
         self.__scalecb.activated.connect(self.edited)
 
     def setParameters(self, params):
-        center = params.get("center", Scale.CenterMean)
-        scale = params.get("scale", Scale.ScaleBySD)
+        center = params.get("center", Scale.CenterMean - 1)
```
Comment (astaric) [inline Orange/widgets/data/owpreprocess.py]: Why - 1?

File: Orange/widgets/data/owpreprocess.py [hunk 9]
```diff
@@ -762,8 +700,8 @@ def __init__(self, parent=None, **kwargs):
...
+        scale = params.get("scale", Scale.ScaleBySD - 1)
```
Comment (astaric) [inline Orange/widgets/data/owpreprocess.py]: Why -1?

File: Orange/widgets/utils/owlearnerwidget.py [hunk 10]
```diff
@@ -194,6 +194,7 @@ def check_data(self):
                 self.valid_data = True
         return self.valid_data
 
+
```
Comment (astaric) [inline Orange/widgets/utils/owlearnerwidget.py]: One line between methods is usually sufficient.

File: Orange/preprocess/fss.py [hunk 11]
```diff
@@ -89,6 +89,7 @@ def __call__(self, data):
                                     data.domain.class_vars, data.domain.metas)
         return data.from_table(domain, data)
 
+
```
Comment (astaric) [inline Orange/preprocess/fss.py]: One empty line is usually enough.

**Comment by astaric** [comment]
Another note, do we state anywhere (docs, github, wiki) that the strings returned by __repr__ should be executable? If not, should we, and where?

File: Orange/base.py [hunk 12]
```diff
@@ -72,14 +74,46 @@ def preprocess(self, data):
...
         return True
 
+    @property
+    def name(self):
+        """Return a short name derived from Learner type name"""
+        try:
+            return self.__name
+        except AttributeError:
+            name = self.__class__.__name__
+            assert name.endswith('Learner'), \
```
Comment (kernc) [inline Orange/base.py]: To enforce some conformity. Now stripped.

File: Orange/classification/logistic_regression.py (hunk 2 above)
Comment (kernc) [inline Orange/classification/logistic_regression.py]: Number of results for
https://www.google.si/search?num=2&q=logit
vs.
https://www.google.si/search?num=2&q=logreg

Been bugging me for a while.

File: Orange/preprocess/continuize.py (hunk 4 above)
Comment (kernc) [inline Orange/preprocess/continuize.py]: Unused.

File: Orange/tests/test_classification.py (hunk 6 above)
Comment (kernc) [inline Orange/tests/test_classification.py]: Seems to be via `Orange.evaluation.AUC` below. I don't know how that worked before, or how Python's imports work, obviously. :confused:

File: Orange/widgets/data/owpreprocess.py (hunk 8 above)
Comment (kernc) [inline Orange/widgets/data/owpreprocess.py]: The enum values are no longer comparable to integers, which IMHO is a good thing. Replaced with two short utility functions.

**Comment by kernc** [comment]
Comments addressed, `__repr__` mentioned. Thanks.

File: Orange/classification/logistic_regression.py (hunk 2 above)
Comment (lanzagar) [inline Orange/classification/logistic_regression.py]: a) the number of results for "trump" is even greater, lets not call it that!
More seriously:
b) the number of results for "logistic regression" is over 5x that of "logit"
Even more seriously:
c) while I don't like nitpicking over terminology (especially when not very well defined), I think logit is more frequently used for the function, and logistic regression for the predictive model (which uses the function). Wikipedia has "logit regression" as an alternative, but I don't think it is used as much. And logit shows an article which has this at the top: "For the basic regression technique that uses the logit function, see logistic regression"

Anyway as long as we are not renaming the class to LogitRegression the name change gets a -1 from me. But, you could argue for a non abbreviated name if that bothers you...

File: Orange/tests/test_classification.py (hunk 6 above)
Comment (astaric) [inline Orange/tests/test_classification.py]: Could you replace with explicit import?

File: Orange/preprocess/continuize.py [hunk 13]
```diff
@@ -8,15 +8,18 @@
...
+                 transform_class=False, **kwargs):
         self.zero_based = zero_based
         self.multinomial_treatment = multinomial_treatment
         self.transform_class = transform_class
 
-        return self if data is None else self(data)
+        # Backward-compat polite notice
+        if isinstance(zero_based, (Table, Domain)) or 'data' in kwargs:
+            raise TypeError('DomainContinuizer no longer accepts data in the '
+                            'constructor. Instea, first make an object instance, '
```
Comment (astaric) [inline Orange/preprocess/continuize.py]: *Instead

File: Orange/preprocess/util.py [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+class _RefuseDataInConstructor:
+    # TODO: drop this as soon as possible
+    def __new__(cls, *args, **kwargs):
+        if (args and isinstance(args[0], Table) or
+                isinstance(kwargs.get('data'), Table)):
+            warnings.warn(
+                '%s no longer accepts data in the constructor. Instead, '
+                'first make an instance, i.e. %s(), then call it with data.'
+                % (cls, cls), OrangeDeprecationWarning, stacklevel=2)
+        return super().__new__(cls)
```
Comment (astaric) [inline Orange/preprocess/util.py]: IMHO DeprecationWarnings are used to point that a feature/method/class is going away and that a developer should modify the code as it will not work with the future versions of the library. Here, a warning is used to point that the code does not work any more.

I would rather have a ValueError raised (with the same message), or keep the old behaviour working until some future version (probably mentioned in the warning).

File: Orange/projection/pca.py [hunk 15]
```diff
@@ -154,8 +154,7 @@ def compute(self, data, pca_space):
         return pca_space[:, self.feature]
 
 
-class RemotePCA:
-    def __new__(cls, data, batch=100, max_iter=100):
+def remotePCA(data, batch=100, max_iter=100):
```
Comment (astaric) [inline Orange/projection/pca.py]: Please update the import in widget as well.
https://github.com/biolab/orange3/blob/master/Orange/widgets/unsupervised/owpca.py#L154

File: Orange/preprocess/util.py (hunk 14 above)
Comment (kernc) [inline Orange/preprocess/util.py]: > or keep the old behaviour working until some future version

That's what I meant. :blush:  Done. Thanks.

File: Orange/projection/pca.py (hunk 15 above)
Comment (kernc) [inline Orange/projection/pca.py]: Reverted because it doesn't play along with Orange remote.

File: Orange/widgets/unsupervised/owpca.py [hunk 16]
```diff
@@ -151,9 +151,9 @@ def start(self):
         else:
             self.address = self.addresstext.text()
             with remote.server(self.address):
-                from Orange.projection.pca import RemotePCA
+                from Orange.projection.pca import remotePCA
```
Comment (astaric) [inline Orange/widgets/unsupervised/owpca.py]: This should be reverted as well

File: Orange/widgets/data/owcontinuize.py [hunk 17]
```diff
@@ -346,17 +347,15 @@ def normalize_by_sd(var, data_or_dist):
     return normalized_var(var, mean, 1 / sd)
 
 
-class DomainContinuizer:
-    def __new__(cls, data=None, zero_based=True,
-                multinomial_treatment=Continuize.Indicators,
-                continuous_treatment=Continuize.Leave,
-                class_treatment=Continuize.Leave):
-        self = super().__new__(cls)
+class DomainContinuizer(Reprable):
```
Comment (astaric) [inline Orange/widgets/data/owcontinuize.py]: Should this class also be backwards compatible (with a warning)?

File: Orange/widgets/unsupervised/owpca.py (hunk 16 above)
Comment (kernc) [inline Orange/widgets/unsupervised/owpca.py]: This creeped in in the wrong place. Thanks.
